### PR TITLE
Part: Make PreviewExtension usable from Python workbenches

### PR DESCRIPTION
**Pull request by kadet1090**
This will allow Python features to use the preview feature without the need of subclassing in C++.

It exposes all of the features in C++ to Python classes via proxies and extensions in a full feature-parity way. Python workbenches now can utilize previews - both simple ones and more complex ones with multiple different shapes.

## Issues
- As part of https://github.com/FreeCAD/FPA-grant-proposals/issues/22

### CODE REVIEW

File: src/Mod/Part/App/PreviewExtension.cpp [hunk 1]
```diff
@@ -24,9 +24,36 @@
...
+namespace Part
+{
+
+template<typename ExtensionT>
+App::DocumentObjectExecReturn* PreviewExtensionPythonT<ExtensionT>::recomputePreview()
+{
+    EXTENSION_PROXY_NOARG(recomputePreview)
+
+    if (!result.isNone() && result.isString()) {
+        return new App::DocumentObjectExecReturn(static_cast<std::string>(Py::String(result)));
```
Comment (tritao) [inline src/Mod/Part/App/PreviewExtension.cpp]: Is this error path actually handled somewhere right now? From what I can tell the callers just leak it if Python returns a string.

File: src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp [hunk 2]
```diff
@@ -196,6 +206,50 @@ void ViewProviderPreviewExtension::extensionOnChanged(const App::Property* prop)
...
+    if (!object) {
+        ViewProviderExtension::extensionUpdateData(prop);
+        return;
+    }
+
+    if (auto* previewExtension = object->getExtensionByType<Part::PreviewExtension>(true)) {
+        if (prop == &previewExtension->PreviewShape) {
+            updatePreview();
+        }
+        else if (isPreviewEnabled()) {
```
Comment (tritao) [inline src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp:236]: Previously PartDesign only queued preview work when the preview had actually been invalidated. Would it make sense to restore that here with a `!previewExtension->isPreviewFresh()` check?

Comment (kadet1090) [reply src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp:236]: I'll need to measure that. The queue and that check aimed to do basically the same thing - prevent unnecessary recomputes. For my understanding it will only prevent scheduling which itself is pretty cheap so it should not make a big difference.

I'll however double check as I removed that part early in the testing and maybe it should indeed stay here.

File: src/Mod/Part/App/PreviewExtension.cpp (hunk 1 above)
Comment (kadet1090) [reply src/Mod/Part/App/PreviewExtension.cpp]: Good catch. I have not thought about it essentially being a leak - in my mind it was returning a value object basically but it ain't java after all... 

It's not used right now, the goal (when designing the original C++ implementation) was so the preview engine can return error state that then could be processed and presented in the UI. I skipped handling it for now but left the interface that would be the long-term goal to not change it later. I'll see if what I can do about that.

File: src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp (hunk 2 above)
Comment (kadet1090) [reply src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp:236]: @tritao I have checked and the method that actually recomputes the preview is called the same amount of times so there is no difference here. I think we should leave the simplified condition here.

File: src/Mod/Part/App/PreviewExtension.cpp (hunk 1 above)
Comment (kadet1090) [reply src/Mod/Part/App/PreviewExtension.cpp]: @tritao Ok, I removed the leak by making the property static and only changing the underlying text so it is shared and does not need to be deleted. I don't think it's the best approach but should be good enough for now. I am however open for better solutions if you have any in mind.

**Review by tritao — APPROVED** [review]

File: src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    Twin="ViewProviderPreviewExtension",
+    TwinPointer="ViewProviderPreviewExtension",
+    Include="Mod/Part/Gui/ViewProviderPreviewExtension.h",
+    FatherInclude="Gui/ViewProviderExtensionPy.h",
+    Namespace="PartGui",
+    FatherNamespace="Gui",
+)
+class ViewProviderPreviewExtension(ViewProviderExtension):
+    """
+    Provides a 3D preview of an upcoming geometry change.
```
Comment (hyarion) [inline src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi:19]: can you add a comment here that it is experimental

Comment (maxwxyz) [reply src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi:19]: do we need a reminder to remove "experimental" is 1.2 if released?x

**Comment by maxwxyz** [comment]
@shaise FYI regarding transparent previews for e.g. sheetmetal

**Comment by shaise** [comment]
yes, on my list.

**Comment by shaise** [comment]
@kadet1090 , any idea when will this be merged? I want to start implementing it with sheetmetal. 
Or if anyone have a working windows binary with this fix, so I can make tests with it.

**Comment by maxwxyz** [comment]
@kadet1090 ping for conflict

**Comment by maxwxyz** [comment]
@kadet1090 mark as ready for review once addressed

**Comment by kadet1090** [comment]
@tritao I reworked the PR to have full feature-parity with what is available in C++ and because of that I removed the experimental markers. I'd appreciate re-review so we can merge it in and expose this to python devs.

**Comment by tritao** [comment]
As usual I've started with an AI audit, it found a couple of issues, and after some back and forth, ended up with the following suggestions:

**1. Replace `mustRecomputePreview()` with a property-aware hook.**

Today:

```cpp
if (isInputProp || mustRecomputePreview()) {
    _isPreviewFresh = false;
}
```

means any normal input property marks the preview stale regardless of a Python/C++ override returning `false`. That defeats the advertised purpose of `mustRecomputePreview()` as an optimization hook.

The current `FreshPreviewFeature` test also sidesteps this because it changes `PreviewShape`, an output property, rather than a normal input such as `Size`.

I think the cleaner long-term API is something like:

```cpp
virtual bool affectsPreview(const App::Property* prop);
```

with `extensionOnChanged()` doing:

```cpp
if (affectsPreview(prop)) {
    invalidatePreview();
}
```

The default should preserve the generic behavior: normal input properties invalidate the preview, while output properties do not. Features can override this when their preview depends on only a subset of their inputs.

For Python I'd pass the property name, consistent with the existing `onChanged(obj, propertyName)` API:

```python
def affectsPreview(self, ext, propertyName):
    if propertyName == "Size":
        return False
    return None  # use the default
```

`None` falling back to the C++ implementation is useful here: proxies only need to special-case the properties they care about instead of reproducing the framework's default rules.

Once a property is declared not to affect the preview, `extensionUpdateData()` should also avoid scheduling a known no-op update when the preview is still fresh.

Please add a regression test where the preview is fresh, `Size` changes, `affectsPreview(..., "Size")` returns `False`, and the preview remains fresh / no recompute is scheduled.

There is also a small pre-existing issue in `EXTENSION_PROXY_ONEARG` that this API would start relying on with the suggested diff below: the `__object__` path constructs a zero-length `Py::Tuple` and then writes item 0. `FeaturePython::onChanged()` correctly constructs a one-element tuple for the equivalent case. Since `affectsPreview()` needs a one-argument proxy call, it makes sense to fix that helper here too.

**2. Empty shape should explicitly clear the preview.**

`updatePreviewShape()` currently returns on `shape.isNull()`, leaving old Coin geometry visible. I think the contract should be:

* valid shape → replace preview
* empty/null shape → clear preview
* rendering failure → report it and leave a deterministic state, preferably cleared

That makes `Part.Shape()` naturally mean “no preview”.

`setupCoinGeometry()` already appears to support this: the existing exception path calls it with `{}` to clear the node. So this should mostly mean allowing an empty shape through to `setupCoinGeometry()` and avoiding `getTransform()` on it.

Please add a regression test that renders a non-empty shape, replaces it with `Part.Shape()`, and verifies the old geometry is gone.

**A couple of small cleanup points while here:**

* Some of the new test comments refer to internal context such as `api-corrections.md` and “Task 9”. I'd make those comments self-contained so they still make sense later without external implementation notes.
* Since the Python docs explicitly tell workbench authors to `ref()` Coin nodes they retain, it would be good to also document the matching ownership/release expectation (`unref()` when that manual reference is no longer needed).

A rough implementation of all of the above could look like:

```diff
diff --git a/src/App/ExtensionPython.h b/src/App/ExtensionPython.h
--- a/src/App/ExtensionPython.h
+++ b/src/App/ExtensionPython.h
@@
 #define EXTENSION_PROXY_ONEARG(function, arg)                                                      \
     EXTENSION_PROXY_FIRST(function)                                                                \
-    Py::Tuple args;                                                                                \
+    Py::Tuple args(1);                                                                             \
     args.setItem(0, arg);                                                                          \
     EXTENSION_PROXY_SECOND(function)                                                               \
     Py::Tuple args(2);                                                                             \
     args.setItem(0, Py::Object(this->getExtensionPyObject(), true));                               \
     args.setItem(1, arg);                                                                          \
     EXTENSION_PROXY_THIRD()

diff --git a/src/Mod/Part/App/PreviewExtension.h b/src/Mod/Part/App/PreviewExtension.h
--- a/src/Mod/Part/App/PreviewExtension.h
+++ b/src/Mod/Part/App/PreviewExtension.h
@@
     /// Marks the preview as stale so the next updatePreview() recomputes it.
     void invalidatePreview();

-    virtual bool mustRecomputePreview();
+    /// Returns whether a change to the given property invalidates the preview.
+    virtual bool affectsPreview(const App::Property* prop);

     PyObject* getExtensionPyObject() override;

@@
 protected:
     App::DocumentObjectExecReturn* recomputePreview() override;
-    bool mustRecomputePreview() override;
+    bool affectsPreview(const App::Property* prop) override;
 };

diff --git a/src/Mod/Part/App/PreviewExtension.cpp b/src/Mod/Part/App/PreviewExtension.cpp
--- a/src/Mod/Part/App/PreviewExtension.cpp
+++ b/src/Mod/Part/App/PreviewExtension.cpp
@@
 template<typename ExtensionT>
-bool PreviewExtensionPythonT<ExtensionT>::mustRecomputePreview()
+bool PreviewExtensionPythonT<ExtensionT>::affectsPreview(const App::Property* prop)
 {
-    EXTENSION_PROXY_NOARG(mustRecomputePreview)
+    const char* name = this->getExtendedObject()->getPropertyName(prop);
+    Py::String propertyName(name ? name : "");
+
+    EXTENSION_PROXY_ONEARG(affectsPreview, propertyName)

     if (result.isBoolean()) {
         return Py::Boolean(result);
     }

-    return ExtensionT::mustRecomputePreview();
+    return ExtensionT::affectsPreview(prop);
 }

@@
-bool Part::PreviewExtension::mustRecomputePreview()
+bool Part::PreviewExtension::affectsPreview(const App::Property* prop)
 {
-    return getExtendedObject()->mustRecompute();
+    return prop && !getExtendedObject()->isOutputProperty(prop);
 }

@@
 void Part::PreviewExtension::extensionOnChanged(const App::Property* prop)
 {
     DocumentObjectExtension::extensionOnChanged(prop);

-    // Invalidate the preview on any input property change. The PreviewShape itself is the
-    // output of recomputePreview, so changing it does not require another recompute.
-    const bool isInputProp = prop != &PreviewShape && !getExtendedObject()->isOutputProperty(prop);
-
-    if (isInputProp || mustRecomputePreview()) {
-        _isPreviewFresh = false;
+    if (affectsPreview(prop)) {
+        invalidatePreview();
     }
 }

diff --git a/src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp b/src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp
--- a/src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp
+++ b/src/Mod/Part/Gui/ViewProviderPreviewExtension.cpp
@@
     if (auto* previewExtension = object->getExtensionByType<Part::PreviewExtension>(true)) {
         if (prop == &previewExtension->PreviewShape) {
             updatePreview();
         }
-        else if (isPreviewEnabled()) {
+        else if (isPreviewEnabled() && !previewExtension->isPreviewFresh()) {
             // Properties can be updated in batches, where some properties trigger other updates.
             // We don't need to compute the preview for intermediate steps. Instead of updating
             // the preview immediately (and potentially doing it multiple times in a row), we
@@
 void ViewProviderPreviewExtension::updatePreviewShape(Part::TopoShape shape, SoPreviewShape* preview)
 {
-    if (shape.isNull() || preview == nullptr) {
+    if (preview == nullptr) {
         return;
     }

     auto vp = freecad_cast<ViewProviderPartExt*>(getExtendedViewProvider());

@@
-    const auto updatePreviewShape = [vp](SoPreviewShape* preview, Part::TopoShape shape) {
+    const auto setupGeometry = [vp](SoPreviewShape* preview, Part::TopoShape shape) {
         ViewProviderPartExt::setupCoinGeometry(
             shape.getShape(),
             preview,
@@
     };

     try {
-        updatePreviewShape(preview, shape);
+        setupGeometry(preview, shape);
+
+        if (shape.isNull()) {
+            return;
+        }
+
         preview->transform.setValue(Base::convertTo<SbMatrix>(shape.getTransform()));
     }
     catch (Standard_Failure& e) {
@@
-        updatePreviewShape(preview, {});
+        setupGeometry(preview, {});
+        return;
     }

diff --git a/tests/src/Mod/Part/App/PreviewExtension.cpp b/tests/src/Mod/Part/App/PreviewExtension.cpp
--- a/tests/src/Mod/Part/App/PreviewExtension.cpp
+++ b/tests/src/Mod/Part/App/PreviewExtension.cpp
@@
     App::PropertyInteger TypeOnlyOutput;

     int recomputePreviewCalls {0};
+    bool ignoreTriggerForPreview {false};

 protected:
     App::DocumentObjectExecReturn* recomputePreview() override
     {
         ++recomputePreviewCalls;
         return App::DocumentObject::StdReturn;
     }
+
+    bool affectsPreview(const App::Property* prop) override
+    {
+        if (ignoreTriggerForPreview && prop == &Trigger) {
+            return false;
+        }
+        return Part::PreviewExtension::affectsPreview(prop);
+    }
 };

@@
 TEST_F(PreviewExtensionTest, changingInputPropertyInvalidates)
 {
     auto* feature = getFeature();
     feature->updatePreview();
@@
     EXPECT_FALSE(feature->isPreviewFresh());
 }

+TEST_F(PreviewExtensionTest, ignoredInputPropertyDoesNotInvalidate)
+{
+    auto* feature = getFeature();
+    feature->updatePreview();
+    ASSERT_TRUE(feature->isPreviewFresh());
+
+    const int callsBefore = feature->recomputePreviewCalls;
+    feature->ignoreTriggerForPreview = true;
+    feature->Trigger.setValue(42);
+
+    EXPECT_TRUE(feature->isPreviewFresh());
+
+    feature->updatePreview();
+    EXPECT_EQ(feature->recomputePreviewCalls, callsBefore);
+}
+
@@
-TEST_F(PreviewExtensionTest, mustRecomputePreviewFollowsMustRecompute)
-{
-    auto* feature = getFeature();
-
-    EXPECT_EQ(feature->mustRecomputePreview(), feature->mustRecompute());
-}

diff --git a/src/Mod/Part/TestPartPreview.py b/src/Mod/Part/TestPartPreview.py
--- a/src/Mod/Part/TestPartPreview.py
+++ b/src/Mod/Part/TestPartPreview.py
@@
-class FreshPreviewFeature(PreviewFeature):
-    """Feature that declares its preview never needs recomputing."""
+class SelectivePreviewFeature(PreviewFeature):
+    """Feature whose preview intentionally ignores Size changes."""

-    def mustRecomputePreview(self, ext):
-        return False
+    def affectsPreview(self, ext, propertyName):
+        if propertyName == "Size":
+            return False
+        return None

@@
-class TestMustRecomputePreviewRouting(unittest.TestCase):
+class TestAffectsPreviewRouting(unittest.TestCase):
@@
-        self.feature = FreshPreviewFeature(self.object)
+        self.feature = SelectivePreviewFeature(self.object)
@@
-    def testReturningFalseKeepsPreviewFresh(self):
+    def testReturningFalseForInputKeepsPreviewFresh(self):
         self.object.updatePreview()
         self.assertTrue(self.object.isPreviewFresh())
+        callsBefore = self.feature.recomputePreviewCalls

-        self.object.touch()
-        self.object.PreviewShape = Part.makeBox(1, 1, 1)
+        self.object.Size = 20.0

         self.assertTrue(self.object.isPreviewFresh())
+        self.object.updatePreview()
+        self.assertEqual(self.feature.recomputePreviewCalls, callsBefore)

@@
     def testUpdatePreviewShapePopulatesGeometry(self):
         from pivy import coin

         node = coin.SoType.fromName("SoPreviewShape").createInstance()
@@
         self.assertGreaterEqual(coordinates.point.getNum(), 8)

+    def testUpdatePreviewShapeWithEmptyShapeClearsGeometry(self):
+        from pivy import coin
+
+        node = coin.SoType.fromName("SoPreviewShape").createInstance()
+        self.object.ViewObject.updatePreviewShape(Part.makeBox(5, 5, 5), node)
+
+        search = coin.SoSearchAction()
+        search.setType(coin.SoCoordinate3.getClassTypeId())
+        search.apply(node)
+        coordinates = search.getPath().getTail()
+
+        self.assertGreaterEqual(coordinates.point.getNum(), 8)
+
+        self.object.ViewObject.updatePreviewShape(Part.Shape(), node)
+
+        self.assertLessEqual(coordinates.point.getNum(), 1)
+
@@
 class TestPreviewNodeAccessBeforeAttach(unittest.TestCase):
-    """Covers addExtension() called after the view provider has already
-    attached, per api-corrections.md: registerExtension() alone does not run
-    extensionAttach(), so the preview nodes can legitimately still be null.
-    """
+    """Covers addExtension() after the view provider has already attached.
+
+    Adding the extension at this point does not call extensionAttach(), so the
+    preview nodes can legitimately still be null.
+    """

@@
 class TestExtensionRegisteredNameUnchanged(unittest.TestCase):
-    """Verifies the biggest risk of Task 9: swapping in the
-    ViewProviderPreviewExtensionPythonT intermediate must not change the
-    registered extension type name, since BIM's ArchWindow.py and every
-    workbench look it up by this exact literal string."""
+    """The Python intermediate must not change the registered extension type
+    name, since callers look it up by this exact literal string."""

diff --git a/src/Mod/Part/previews.dox b/src/Mod/Part/previews.dox
--- a/src/Mod/Part/previews.dox
+++ b/src/Mod/Part/previews.dox
@@
- * properties cannot affect the preview can say so through
- * `mustRecomputePreview()`, and skip the work entirely.
+ * properties cannot affect the preview can say so through
+ * `affectsPreview()`, and skip the work entirely.
@@
- * - `mustRecomputePreview` decides when that work is worth doing, and is how a
- *   feature avoids recomputing for properties its preview does not depend on.
+ * - `affectsPreview` decides whether a particular property change invalidates
+ *   the preview. Python hooks receive the property's name.
@@
- * Coin nodes a Python workbench keeps between refreshes need an explicit
- * reference; a Python variable alone does not keep them alive.
+ * Coin nodes a Python workbench keeps between refreshes need an explicit
+ * reference; a Python variable alone does not keep them alive. A manual
+ * `ref()` should be paired with `unref()` when that reference is no longer
+ * needed.

diff --git a/src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi b/src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi
--- a/src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi
+++ b/src/Mod/Part/Gui/ViewProviderPreviewExtension.pyi
@@
     A Coin node held in a Python attribute across calls - e.g. shared between
     updatePreview() invocations - is not kept alive by that reference alone;
     unlike Gui::CoinPtr on the C++ side, a SWIG pointer does not ref() the
-    underlying SoBase. Call node.ref() yourself, or the node may be freed the
-    moment its last scene graph parent is removed, crashing on next use.
+    underlying SoBase. Call node.ref() while retaining such a node, and pair
+    that reference with node.unref() when it is no longer needed.
```

**Comment by kadet1090** [comment]
1. Is out-of scope for such PR. While I'm not against doing something like that - this is a change that is more C++  and I'd also preserve the mostRecomputePreview call so it can be implemented not only in terms of properties. Current solution also mimics the current API for mustRecompute and I don't really want to change it - that said - I think that reframing default implementation of `mustRecomputePreview` in terms of such function would be good - but as follow-up.

2. Ah, yes, that's also a problem with C+. One that has bitten me with geometry selector recently.  I'll put a separate PR for it - also out of scope for this PR.

Good point about comments - they leaked and I have not noticed it.

**Comment by tritao** [comment]
In that case, I think the main thing we should resolve in this PR is the mismatch between the code and the docs.

Right now the docs say that `mustRecomputePreview()` can be used to avoid recomputing the preview for input properties it does not depend on, but with:

```cpp
if (isInputProp || mustRecomputePreview()) {
    _isPreviewFresh = false;
}
```

any normal input property already makes the preview stale, so returning false cannot actually prevent that.

So I think either the docs should be adjusted to describe the narrower behavior the current API really provides, or the API should be changed so that the documented optimization is possible.

But otherwise this looks good, will be great to start using this in BIM.

**Comment by kadet1090** [comment]
@tritao instead of updating docs I restored the original call that only relies on mustRecomputePreview - I was chasing some bug before and that was one of the things I played with, but it shouldn't be needed.

**Comment by tritao** [comment]
Looks good, only one small nit remains: the documentation says an empty shape makes the preview disappear, which as we discussed is not really the case yet.

**Review by tritao — APPROVED** [review]

**Comment by kadet1090** [comment]
@mnesarco @shaise - can you take a look at it (especially the docs) and tell me if that seems ok?

File: src/Mod/Part/previews.dox [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+ *
+ * @section preview_pitfalls Things worth knowing
+ *
+ * The preview is refreshed automatically when the published shape changes.  An
+ * arrangement that depends on anything else — a count, a set of transformations
+ * — has to ask for a refresh when that changes.
+ *
+ * A failure inside a Python hook is reported rather than propagated, which
+ * leaves the previous preview on screen.  A hook that cannot compute its
+ * geometry should publish an empty shape, so the preview disappears instead of
```
Comment (mnesarco) [inline src/Mod/Part/previews.dox:299]: Apparently this is not true yet. Empty shape does not remove the preview as pointed out by @tritao

File: src/Mod/Part/previews.dox [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+ *    volume a cut removes rather than what remains,
+ * 3. the ability to show any number of shapes at once, each with its own
+ *    appearance, and
+ * 4. the ability to draw one shape many times, as a pattern does for its
+ *    instances.
+ *
+ * The first point is what a feature gets for free.  The rest is what a feature
+ * can reach for when the default is not expressive enough.
+ *
+ * The framework is available to C++ and to Python workbenches on equal terms.
```
Comment (mnesarco) [inline src/Mod/Part/previews.dox:26]: Except style parameters.

**Review by mnesarco — APPROVED** [review]
Only to comments. but LGTM.

File: src/Mod/Part/previews.dox (hunk 5 above)
Comment (kadet1090) [reply src/Mod/Part/previews.dox:26]: I'm working on that right now - expect PR with FEP perhaps soon :)

File: src/Mod/Part/previews.dox (hunk 4 above)
Comment (kadet1090) [reply src/Mod/Part/previews.dox:299]: yes, I'll push PR fixing it today - docs describe intended state, that one is simply a bug

**Review by maxwxyz — APPROVED** [review]
Tested

**Comment by shaise** [comment]
is there a documentation or a sample code how to use it?

**Comment by furgo16** [comment]
@shaise See https://github.com/FreeCAD/FreeCAD/pull/29642#pullrequestreview-4179374571 - I made a small summary of the API, and the changes in the PR should give you an overview of how it's used. Thanks @kadet1090 for this work!

**Comment by shaise** [comment]
Great! Thanks!
